korma/src/commonMain/kotlin/com/soywiz/korma/geom/Vector3D.kt

```diff
@@ -5,95 +5,190 @@ import com.soywiz.korma.interpolation.interpolate
 import com.soywiz.korma.math.*
 import kotlin.math.*
 
-class Vector3D {
-    val data = floatArrayOf(0f, 0f, 0f, 1f)
+inline fun vectorLength(x: Float, y: Float, z: Float) = sqrt(vectorLengthSquared(x, y, z))
+inline fun vectorLengthSquared(x: Float, y: Float, z: Float) = (x * x) + (y * y) + (z * z)
 
+class Vector3D constructor(
+    val data: FloatArray
+) {
     var x: Float get() = data[0]; set(value) = run { data[0] = value }
     var y: Float get() = data[1]; set(value) = run { data[1] = value }
     var z: Float get() = data[2]; set(value) = run { data[2] = value }
-    var w: Float get() = data[3]; set(value) = run { data[3] = value }
 
-    val lengthSquared: Float get() = (x * x) + (y * y) + (z * z) + (w * w)
+    val lengthSquared: Float get() = (x * x) + (y * y) + (z * z)
     val length: Float get() = sqrt(lengthSquared)
 
-    val length3Squared: Float get() = (x * x) + (y * y) + (z * z)
-    val length3: Float get() = sqrt(length3Squared)
+    constructor(x: Float, y: Float, z: Float): this(floatArrayOf(x, y, z))
+    constructor(func: (index: Int) -> Float): this(func(0), func(1), func(2))
```

Comment:
This could be (to prevent allocating the lambda):

```
companion object {
    inline operator fun invoke(func: (index: Int) -> Float) = FloatArray(4) { func(it) }
}
```